### PR TITLE
adding ingress

### PR DESCRIPTION
**Pull request by ofer-velich**

### CODE REVIEW

File: apollo-backend/src/main/java/io/logz/apollo/kubernetes/ApolloToKubernetes.java [hunk 1]
```diff
@@ -135,14 +144,43 @@ public Service getKubernetesService() throws ApolloParseException {
...
-                    deploymentTransformers.size(), apolloDeployment.getId());
+                    serviceTransformers.size(), apolloDeployment.getId());
             serviceTransformers.forEach(transformer ->
                     transformer.transform(service, apolloDeployment, apolloService, apolloEnvironment, apolloDeployableVersion));
 
             return service;
 
         } catch (IOException e) {
-            throw new ApolloParseException("Could not parse deployment YAML from DB", e);
+            throw new ApolloParseException("Could not parse ingress YAML from DB", e);
```
Comment (ShaharGotshtat) [inline apollo-backend/src/main/java/io/logz/apollo/kubernetes/ApolloToKubernetes.java]: Why? isn't it still possible that the problem is in the deployment YAML?

File: apollo-backend/src/main/java/io/logz/apollo/transformers/ingress/BaseIngressTransformer.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package io.logz.apollo.transformers.ingress;
+
+import io.fabric8.kubernetes.api.model.extensions.Ingress;
+
+/**
+ * Created by roiravhon on 1/31/17.
```
Comment (ShaharGotshtat) [inline apollo-backend/src/main/java/io/logz/apollo/transformers/ingress/BaseIngressTransformer.java]: This is not true :)

File: apollo-backend/src/main/java/io/logz/apollo/transformers/ingress/IngressLabelTransformer.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    public Ingress transform(Ingress ingress,
+                             io.logz.apollo.models.Deployment apolloDeployment,
+                             io.logz.apollo.models.Service apolloService,
+                             io.logz.apollo.models.Environment apolloEnvironment) {
+
+        Map<String, String> desiredLabels = ImmutableMap.<String, String> builder()
+                .put("environment", LabelsNormalizer.normalize(apolloEnvironment.getName()))
+                .put("geo_region", LabelsNormalizer.normalize(apolloEnvironment.getGeoRegion()))
+                .put("apollo_unique_identifier", ApolloToKubernetes.getApolloServiceUniqueIdentifier(apolloEnvironment,
+                        apolloService, Optional.empty()))
```
Comment (ShaharGotshtat) [inline apollo-backend/src/main/java/io/logz/apollo/transformers/ingress/IngressLabelTransformer.java]: What if there is a group name? This unique identifier would be wrong...

File: apollo-backend/src/test/java/io/logz/apollo/TransformersTest.java [hunk 4]
```diff
@@ -126,6 +126,23 @@ public void testServiceLabelsTransformer() throws ApolloParseException {
...
+        ApolloToKubernetes apolloToKubernetes;
+
+        String sampleLabelFromTransformer = "environment";
+
+        realDeploymentGenerator = new RealDeploymentGenerator("image", "key", "value", 0);
+        apolloToKubernetes = createForDeployment(realDeploymentGenerator.getDeployment());
+        assertIngressLabelExists(apolloToKubernetes.getKubernetesIngress(), realDeploymentGenerator.getDefaultLabelKey(), realDeploymentGenerator.getDefaultLabelValue());
+
+        realDeploymentGenerator = new RealDeploymentGenerator("image", "key", "value", 0);
+        apolloToKubernetes = createForDeployment(realDeploymentGenerator.getDeployment());
```
Comment (ShaharGotshtat) [inline apollo-backend/src/test/java/io/logz/apollo/TransformersTest.java]: Can't you remove these 2 lines and just assert again?

File: ui/app/views/deployments/new.html [hunk 5]
```diff
@@ -248,6 +248,25 @@ <h4 class="modal-title">Kubernetes Service YAML</h4>
...
+    <div class="modal fade" id="show-kubernetes-ingress-yaml" tabindex="-1" role="dialog" aria-hidden="true" style="display: none;">
+        <div class="modal-dialog">
+            <div class="modal-content">
+                <div class="modal-header">
+                    <button type="button" class="close" data-dismiss="modal" aria-hidden="true">×</button>
+                    <h4 class="modal-title">Kubernetes Ingress YAML</h4>
+                </div>
+                <div class="modal-body">
+                    <pre class="apollo-logs">
+{{serviceSelected.ingressYaml}}
```
Comment (ShaharGotshtat) [inline ui/app/views/deployments/new.html]: Indentation.

**Comment by ShaharGotshtat** [comment]
Cool feature and nice work!
The comment about the group name and the unique identifier is the only important one :)

File: apollo-backend/src/main/java/io/logz/apollo/transformers/ingress/IngressLabelTransformer.java (hunk 3 above)
Comment (ofer-velich) [reply apollo-backend/src/main/java/io/logz/apollo/transformers/ingress/IngressLabelTransformer.java]: not sure i follow, why the group name should effect the ingress rule unique identifier ?

**Pull request merged**